File: ios/ParkingKok/Detection/Domain/CandidateHistoryEntry.swift
import Foundation

/// One resolved candidate, as docs/10 §7b's bell lists it.
///
/// ### What it may carry, and what it must not
/// docs/05 §10a fixes the three fields exactly: "An entry keeps the raised-at time, the
/// outcome, and for a confirmed one the record id. Not the location." There is no
/// coordinate, no address and no floor here — the floor on a `저장됨` row is read from the
/// parking record `recordId` points at, which is the only copy of it that exists.
///
/// That is not merely tidiness: §7b's list is the same surface a day later, and docs/09
/// keeps location off the notification it mirrors.
struct CandidateHistoryEntry: Sendable, Equatable, Codable, Identifiable {
    /// The `candidateId`. Also the dedup key — see `CandidateHistoryStoring.append`.
    let id: UUID
    /// When the engine raised the guess, not when it was answered. §7b's right-hand
    /// column shows the moment the phone buzzed, which is what the user is recognising.
    let raisedAt: Date
    let outcome: CandidateOutcome
    /// The `ParkingSession` a confirmation became, so the row can open it and show the
    /// floor. `nil` for every other outcome.
    let recordId: UUID?

    init(id: UUID, raisedAt: Date, outcome: CandidateOutcome, recordId: UUID? = nil) {
        self.id = id
        self.raisedAt = raisedAt
        self.outcome = outcome
        // A record id on a rejection or an expiry would be a record that was never
        // written; the type refuses to hold one rather than leaving it to each call site.
        self.recordId = outcome == .confirmed ? recordId : nil
    }

    /// The entry a resolved candidate becomes.
    init(candidate: ParkingCandidate, outcome: CandidateOutcome, recordId: UUID? = nil) {
        self.init(id: candidate.id, raisedAt: candidate.detectedAt, outcome: outcome, recordId: recordId)
    }
}

File: android/app/src/main/java/com/parkingpin/app/domain/detection/CandidateHistoryEntry.kt
package com.parkingpin.app.domain.detection

import kotlinx.serialization.Serializable

/**
 * What became of a candidate (docs/10_DESIGN_UX_SPEC.md §7b "What is in it").
 *
 * Three outcomes and no others. A candidate the user never answered reads the same
 * whether it ran out its forty-five minutes, was superseded by the next trip, or was
 * retired when the car link came back: nobody answered it, and §7b gives that one word.
 */
@Serializable
enum class CandidateOutcome {
    /** Became a parking record. The row names the floor it became. */
    CONFIRMED,

    /** The user said `주차 아님`. */
    REJECTED,

    /** Nobody answered before it went away. */
    EXPIRED,
}

/**
 * One line of the notification history the bell opens
 * (docs/10_DESIGN_UX_SPEC.md §7b, docs/05_PARKING_DETECTION_ENGINE.md §10a "History").
 *
 * ### Why it is a separate list
 * §10a is explicit that the live candidate slot keeps holding at most one and that this
 * is an append-only list beside it, "because the two answer different questions and
 * giving the slot a second job is how it would end up holding two live candidates by
 * accident".
 *
 * ### What it cannot hold
 * The raised-at time, the outcome, and for a confirmed one the id of the record it
 * became. **No coordinate, no address.** §10a keeps location off this surface and §7b
 * repeats it, so there is deliberately no field here that could carry one — the rule is
 * structural rather than something a future caller has to remember. The floor a
 * confirmed row shows is read from the record at display time, for the same reason.
 */
@Serializable
data class CandidateHistoryEntry(
    /** The candidate this line is about. Unique, so it doubles as the list key. */
    val candidateId: String,
    /** When the candidate was raised — [ParkingCandidate.detectedAtMillis]. */
    val raisedAtMillis: Long,
    val outcome: CandidateOutcome,
    /** The record a [CandidateOutcome.CONFIRMED] entry became. Null for the other two. */
    val recordId: String? = null,
) {
    companion object {
        /**
         * §7b Retention: the last 30, and older ones fall off.
         *
         * "A user looking further back wants the parking history, which is a different
         * screen and already keeps everything."
         */
        const val MAX_ENTRIES: Int = 30
    }
}
